rfcs/038-unify-shared-dependencies.md

```diff
@@ -0,0 +1,126 @@
+
+This RFC serves as a platform to figure out the best approach for unifying shared dependencies and contains preliminary proposal for how that could be solved.
+
+## Motivation
+
+Following original RFC:
+
+```
+Dependency Management: Python packages are pretty much the same for each proxy, which makes it hard to sync across all the above repositories. As a result most of the time, each proxy is running its own version of the dependency.
+A few examples:
+
+| Package             | amundsenmetadata   | amundsenfrontend  |
+| --------------      |:------------------:| -----------------:|
+| amundsen-common     | >=0.8.1            | ==0.6.0           |
+| flake8              | ==3.5.0            | ==3.8.4           |
+| flake8-tidy-imports | ==1.1.0            | ==4.2.1           |
+
+and many more... This change will put all the packages in just one place, making practically to keep dependencies similar accross verisons, simply because it will be possible to open a single PR that updates multiple packages. Practically speaking this will improve the status quo slightly.
+```
+
+Most common pain-points so far:
+
+- mismatch between version of amundsen-common in frontend/metadata/search
+- mismatch between version of marshmallow in frontend/metadata/search 
+
+## Guide-level Explanation (aka Product Details)
+
+n/a
+
+## UI/UX-level Explanation
+
+n/a
+
+## Reference-level Explanation (aka Technical Details)
+
+Following steps are proposed to achieve goal of this RFC. 
+
+Multiple times you might find the term `subdirectory`, which refers to following folders:
+- common
+- databuilder
+- frontend
+- metadata
+- search
+
+Steps:
+
+0. Undertake new approach where all dependencies land in `install_requires` or `extras_require` through reading in appropriate requirements* files (except databuilder due to it's complexity)
+1. Split `requirements.txt` for each subdirectory into `requirements.txt` and `requirements-dev.txt` (there is no need to install flake8 when doing for example `pip install amundsen-metadata` which is currently the case)
+2. Identify libraries that are shared across all (or majority - see `Drawbacks`) of subprojects
+3. Create `requiremnts-common.txt` file in root folder of amundsen repo containing pinned dependencies for shared libraries
+4. Remove libraries listed in `requirements-common.txt` from `requirements.txt` of each subdirectory
+5. Create `requirements-dev.txt` file in root folder of amundsen repo containing pinned dependencies for shared dev libraries (mypy, flake8, etc.)
+6. Remove libraries listed in `requirements-dev.txt` from `requirements.txt` of each subdirectory
+7. Create symbolic link to `requirements-common.txt` and `requirements-dev.txt` in each subdirectory
+8. Shift context of building docker images to root folder (otherwise we cannot share `requirements-common.txt` and `requirements-dev.txt` symlinks across subdirectories)
```

Comment:
I think the idea of shifting the build context up one directory makes sense, it also gives you the ability to copy in the `amundsen common` that is on the same brach (not in pypi) and copy it into each image. This could also give the benefit that we wouldn't need to add new versions of `amundsen-common` to pypi.

Reply:
How would we go about pypi installation of amundsen services if common is not a pypi pckage? I think not all users use pre-built images. 

Moreover, we have introduced apache atlas utils recently as a part of common and will be relying on them for both databuilder and metadata.

Reply:
Is the pypi installation of search, metadata and frontend really that common of use-case? I get that we wouldn't want to pull the plug on that user group, and by no means am I against keeping it in pypi and just overriding it during the build process, but it just feels like one of those areas where we _shouldn't need_ to have the common directory in pypi any more since it's now in the same branch - mostly thinking out loud here.